snap/quota/quota.go

```diff
@@ -832,6 +852,40 @@ func (grp *Group) NewSubGroup(name string, resourceLimits Resources) (*Group, er
 	return subGrp, nil
 }
 
+// VerifyNesting takes a group and verifies that it satisfies the following conditions:
+// 1. That if any parent is mixed (has both snaps and sub-groups), it must be the immediate
+//    parent group.
+// 2. If the group itself is mixed, that it has only one level of sub-grouping.
+func (grp *Group) VerifyNesting() error {
+	// A parent group is only allowed to contain a mixture of snaps
+	// and sub-groups if it's a direct parent. Introducing this limitation
+	// will not affect anything as we didn't allow mixing snaps and sub-groups
+	// prior to this change.
+	i := grp.parentGroup
+	for i != nil {
+		if len(i.Snaps) > 0 && len(i.SubGroups) > 0 {
+			// then the group must be a direct parent
+			if grp.parentGroup != i {
+				return fmt.Errorf("group %q is invalid: only one level of sub-groups are allowed for groups mixed with snaps and sub-groups",
+					grp.Name)
+			}
+		}
+		i = i.parentGroup
+	}
+
+	// Now we verify sub-groups, make sure that we are not mixing
+	// snaps and sub-groups with depths deeper than 1
+	if len(grp.Snaps) > 0 && len(grp.SubGroups) > 0 {
```

Comment:
is this check enough? there could snaps deeper into hierarchy?  if the updates are done at different points of an initially  empty hierarchy